<h1> Analiza transportnih podatkov </h1>

<h2> Pregled in urejanje podatkov</h2>

In [380]:
# naložimo Pandas 
import pandas as pd
import numpy as np
# import os.path

# %matplotlib notebook
# Se ne bom poglabljal v grafične predstavitve podatkov.

#Uvoz podatkov-----------
pathData = "data_set - transported.csv"
data = pd.read_csv(pathData)
#------------------------

# Izpisovanje -----------
pd.options.display.max_rows = 30
pd.options.display.max_columns = len(data.columns)
#------------------------


# Številske stolpce Str -> Int
numberColumns = ['Deliveries', 'Attachments', 'Unique destinations']

for i in numberColumns:
    data[i] = data[i].astype(int)

# Weight stolpec Str -> float

data['Weight'] = data['Weight'].apply(lambda x: float(x.replace(',','')))

#------------------------

# data.columns = data.columns.str.replace(' ', '')


In [381]:
data

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    Loading date Loading period Unloading date Unloading period  ZIP - Origin  \
0     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
1     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
2     2022-09-01  00:00 - 23:59     2022-09-05    07:30 - 12:00         52333   
3     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
4     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
..           ...            ...            ...              ...           ...   
386   2022-08-26  00:00 - 00:00     2022-08-28    00:00 - 00:00          3273   
387   2022-08-25  00:00 - 00:00     2022-08-25    00:00 - 00:00          3273   
388   2022-08-24  00:00 - 00:00     2022-08-26    00:00 - 00:00          3273   
389   2022-08-09  00:00 - 00:00     2022-08-10    00:00 - 00:00          3273   
390   2022-08-03  00:00 - 00:00     2022-08-04    00:00 - 00:00          3273   

    City - Origin  Region - Origin Country - Origin  Unique destinations  \
0        Potpican              NaN               HR                    1   
1        Potpican              NaN               HR                    1   
2        Potpican              NaN               HR                    1   
3        Potpican              NaN               HR                    1   
4        Potpican              NaN               HR                    1   
..            ...              ...              ...                  ...   
386    HALMAJUGRA              NaN               HU                    1   
387    HALMAJUGRA              NaN               HU                    1   
388    HALMAJUGRA              NaN               HU                    2   
389    HALMAJUGRA              NaN               HU                    1   
390    HALMAJUGRA              NaN               HU                    1   

     ZIP - Destination           City - Destination  Region - Destination  \
0                15068            Pozzolo Formigaro                   NaN   
1                64010                     Ancarano                   NaN   
2                67100               Poggio Picenze                   NaN   
3                37060  RONCANOVA DI GAZZO VERONESE                   NaN   
4                20865                       USMATE                   NaN   
..                 ...                          ...                   ...   
386              31216                    Antunovac                   NaN   
387              31216                    Antunovac                   NaN   
388              32270                      Zupanja                   NaN   
389              34000                       Požega                   NaN   
390              34000                       Pozega                   NaN   

    Country - Destination  Quote request type  Quote request  \
0                      IT                 NaN            0.0   
1                      IT                 NaN            0.0   
2                      IT                 NaN            0.0   
3             

<h3> Različne vrednosti </h3>

<p>Pogledamo koliko je različnih vrednosti za posamezni stolpec. </p>

In [382]:
#Koliko je različnih vrednosti za posamezen stolpec?

uniqueDataDic = {}
uniqueDataDicCount = {}

for type in data.columns[1:]:
    x = data[type].unique().tolist()
    uniqueDataDic[type] = x
    uniqueDataDicCount[type] = [len(x)]

uniqueDataCount = pd.DataFrame.from_dict(uniqueDataDicCount)

uniqueDataCount

ETA status  ETA type  Difference with ETA  Deliveries  Attachments  \
0           3         2                    6           3            3   

   Loading date  Loading period  Unloading date  Unloading period  \
0            59              19              64                11   

   ZIP - Origin  City - Origin  Region - Origin  Country - Origin  \
0             3              3                1                 3   

   Unique destinations  ZIP - Destination  City - Destination  \
0                    3                139                 145   

   Region - Destination  Country - Destination  Quote request type  \
0                     1                      7                   1   

   Quote request  Quote request currency  Quote request comment  \
0              1                       1                      1   

   Vehicle (Requirement)  Weight  Weight unit  
0                      5     165            1

In [383]:
#Poglejmo neočitne različne vnose po stolpcih

notObviousData = ['ETA status','ETA type','Difference with ETA','Deliveries','Attachments','City - Origin','Country - Origin','Country - Destination','Unique destinations','Vehicle (Requirement)']

#City - Destination -> preveč različnih

for type in notObviousData:
    print(f"{type} : {uniqueDataDic[type]}")

ETA status : ['None', 'On time', 'Delayed']
ETA type : [nan, 'Destination']
Difference with ETA : [' ', '-01:00:43', '-00:09:14', '-00:06:20', '-00:15:03', '00:00:58']
Deliveries : [1, 3, 2]
Attachments : [0, 1, 2]
City - Origin : ['Potpican', 'Caparroso', 'HALMAJUGRA']
Country - Origin : ['HR', 'ES', 'HU']
Country - Destination : ['IT', 'BA', 'HU', 'HR', 'SI', 'RS', 'AT']
Unique destinations : [1, 3, 2]
Vehicle (Requirement) : ['Truck 100 m3 NO', 'Truck 90 m3', 'Megatrailer 95 m3', 'Truck 120 m3 NO', 'HU 24T one-way max 2 stops (5000)']


<p> Že tukaj lahko vidimo neke osnovne podatke o delovanju podjetja.</p>


<h3>Datum</h3>

<p>Uredimo datum. Dani zapis %Y-%m-%d ločimo, dodamo stolpce za mesec in dan.</p>

In [384]:
# Iz datuma dodamo določimo mesec, dan


#Loading date -> v datetime zapis
data['Loading date'] = pd.to_datetime(data['Loading date'],format='%Y-%m-%d')
data['Unloading date'] = pd.to_datetime(data['Unloading date'],format='%Y-%m-%d')

#Dodajanje stolpcev day, month
data['Loading month'] = data['Loading date'].dt.month
data['Loading day'] = data['Loading date'].dt.day

data['Unloading month'] = data['Unloading date'].dt.month
data['Unloading day'] = data['Unloading date'].dt.day


#Časovno trajanje poti

data['Duration (days)'] = (data['Unloading date'] - data['Loading date']).dt.days

data

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    Loading date Loading period Unloading date Unloading period  ZIP - Origin  \
0     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
1     2022-09-01  00:00 - 23:59     2022-09-05    08:00 - 17:00         52333   
2     2022-09-01  00:00 - 23:59     2022-09-05    07:30 - 12:00         52333   
3     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
4     2022-09-01  00:00 - 23:59     2022-09-02    08:00 - 16:00         52333   
..           ...            ...            ...              ...           ...   
386   2022-08-26  00:00 - 00:00     2022-08-28    00:00 - 00:00          3273   
387   2022-08-25  00:00 - 00:00     2022-08-25    00:00 - 00:00          3273   
388   2022-08-24  00:00 - 00:00     2022-08-26    00:00 - 00:00          3273   
389   2022-08-09  00:00 - 00:00     2022-08-10    00:00 - 00:00          3273   
390   2022-08-03  00:00 - 00:00     2022-08-04    00:00 - 00:00          3273   

    City - Origin  Region - Origin  ... Country - Destination  \
0        Potpican              NaN  ...                    IT   
1        Potpican              NaN  ...                    IT   
2        Potpican              NaN  ...                    IT   
3        Potpican              NaN  ...                    IT   
4        Potpican              NaN  ...                    IT   
..            ...              ...  ...                   ...   
386    HALMAJUGRA              NaN  ...                    HR   
387    HALMAJUGRA              NaN  ...                    HR   
388    HALMAJUGRA              NaN  ...                    HR   
389    HALMAJUGRA              NaN  ...                    HR   
390    HALMAJUGRA              NaN  ...                    HR   

     Quote request type  Quote request Quote request currency  \
0                   NaN            0.0                    NaN   
1                   NaN            0.0                    NaN   
2                   NaN            0.0                    NaN   
3                   NaN            0.0                    NaN   
4                   NaN            0.0                    NaN   
..                  ...            ...                    ...   
386                 NaN            0.0                    NaN   
387                 NaN            0.0                    NaN   
388                 NaN            0.0                    NaN   
389                 NaN            0.0                    NaN   
390                 NaN            0.0                    NaN   

     Quote request comment              Vehicle (Requirement)    Weight  \
0                      NaN                    Truck 100 m3 NO   7707.85   
1                      NaN                    Truck 100 m3 NO   8843.54   
2                      NaN                    Truck 100 m3 NO   6134.11   
3                      NaN                    Truck 100 m3 NO   5742.88   
4                      NaN                    Truck 100 m3 NO   9327.65   
..                     ...                                ...       ...   
386                  

In [385]:
#Za katere mesece so podatki

type = 'Loading month'
print( f"Loading month : {data[type].unique().tolist()}")

Loading month : [9, 8, 7, 6, 5]


<h3> Odstranjevanje nepotrebnih podatkov</h3>
 
<p>Poglejmo le uporabne podatke. Odstranimo notUseful stolpce. Glej kodo.</p>

<p>Spodaj so podani urejeni podatki.</p>

In [386]:
notUseful = ['Loading date', 'Unloading date', 'Loading period', 'Unloading period', 'ZIP - Origin', 'Region - Origin', 'ZIP - Destination', 'Region - Destination', 'Quote request type', 'Quote request', 'Quote request currency', 'Quote request comment']

useful = [i for i in data.columns.to_list() if i not in notUseful]

sortedData = data.loc[:, useful]

sortedData

ID ETA status ETA type Difference with ETA  Deliveries  Attachments  \
0      1       None      NaN                               1            0   
1      2       None      NaN                               1            0   
2      3       None      NaN                               1            0   
3      4       None      NaN                               1            0   
4      5       None      NaN                               1            0   
..   ...        ...      ...                 ...         ...          ...   
386  387       None      NaN                               1            0   
387  388       None      NaN                               1            0   
388  389       None      NaN                               2            0   
389  390       None      NaN                               1            0   
390  391       None      NaN                               1            0   

    City - Origin Country - Origin  Unique destinations  \
0        Potpican               HR                    1   
1        Potpican               HR                    1   
2        Potpican               HR                    1   
3        Potpican               HR                    1   
4        Potpican               HR                    1   
..            ...              ...                  ...   
386    HALMAJUGRA               HU                    1   
387    HALMAJUGRA               HU                    1   
388    HALMAJUGRA               HU                    2   
389    HALMAJUGRA               HU                    1   
390    HALMAJUGRA               HU                    1   

              City - Destination Country - Destination  \
0              Pozzolo Formigaro                    IT   
1                       Ancarano                    IT   
2                 Poggio Picenze                    IT   
3    RONCANOVA DI GAZZO VERONESE                    IT   
4                         USMATE                    IT   
..                           ...                   ...   
386                    Antunovac                    HR   
387                    Antunovac                    HR   
388                      Zupanja                    HR   
389                       Požega                    HR   
390                       Pozega                    HR   

                 Vehicle (Requirement)    Weight Weight unit  Loading month  \
0                      Truck 100 m3 NO   7707.85          kg              9   
1                      Truck 100 m3 NO   8843.54          kg              9   
2                      Truck 100 m3 NO   6134.11          kg              9   
3                      Truck 100 m3 NO   5742.88          kg              9   
4                      Truck 100 m3 NO   9327.65          kg              9   
..                                 ...       ...         ...            ...   
386  HU 24T one-way max 2 stops (5000)  20197.80          kg              8   
387  HU 24T one-way max 2 stops (5000)  22157.80          kg              8   
388  HU 24T one-way max 2 stops (5000)  19628.00          kg              8   
389  HU 24T one-way max 2 stops (5000)  20785.80          kg              8   
390  HU 24T one-way max 2 stops (5000)  20785.80          kg              8   

     Loading day  Unloading month  Unloading day  Duration (days)  
0              1                9              5                4  
1              1                9              5                4  
2              1                9              5                4  
3              1                9              2                1  
4              1                9              2                1  
..           ...              ...            ...              ...  
386           26                8             28                2  
387           25                8             25                0  
388           24                8             26                2  
389            9                8             

<h2> Mesečna analiza</h2>

In [387]:
#Mesečna analiza

months = sortedData["Loading month"].unique().tolist()

dataMonthsDic = {'Months' : months}

sumWeight = []
sumDeliveries = []
countInput = []

for month in months:
    countInput.append(len(sortedData[data['Loading month'] == month]))
    sumWeight.append(sortedData.loc[data['Loading month'] == month, 'Weight'].sum())
    sumDeliveries.append(sortedData.loc[data['Loading month'] == month, 'Deliveries'].sum())

dataMonthsDic['All trips'] = countInput
dataMonthsDic['Sum Weight'] = sumWeight
dataMonthsDic['Sum Deliveries'] = sumDeliveries


dataMonths = pd.DataFrame.from_dict(dataMonthsDic)

dataMonths


Months  All trips  Sum Weight  Sum Deliveries
0       9          5    37756.03               5
1       8        112   966294.18             121
2       7        137  1118766.29             143
3       6        135  1080662.47             143
4       5          2    20187.04               2

<p> Pregled začetnih lokacij po posameznih mesecih.</p>

In [388]:
sortedData.groupby(['Loading month', 'City - Origin'])['ID'].count()

Loading month  City - Origin
5              Potpican           2
6              Caparroso          1
               Potpican         134
7              Caparroso          2
               Potpican         135
8              HALMAJUGRA         6
               Potpican         106
9              Potpican           5
Name: ID, dtype: int64

<h2>Lokacijska analiza</h2>


<h3>Začetna lokacija</h3>

<p>Spodaj je tabela vsote teže in števila izvozov iz posameznega mesta.</p>

In [389]:
# Izvoz 

sortedData.groupby('City - Origin')['Weight'].agg(['sum','count'])

sum  count
City - Origin                   
Caparroso        28632.89      3
HALMAJUGRA      124341.00      6
Potpican       3070692.12    382

<h3>Končna lokacija</h3>

<p>Tabela podatkov grupiranih po končni destinaciji s številom prevozov in vsoto teže.</p>

In [390]:
#uvoz

sortedData.groupby('City - Destination')['Weight'].agg(['sum','count']).sort_values(by=['count', 'sum'])

sum  count
City - Destination                  
Borgo Valsugana       3216.96      1
MATERA                4645.38      1
RUFFANO               4708.32      1
INDUNO OLONA          5078.40      1
Dugopolje             5120.64      1
...                       ...    ...
Ancarano            117254.99     14
Michelhausen        201809.30     20
Nichelino           180195.40     23
CASTEL MAGGIORE     212217.84     27
Felsőlajos          519604.37     52

[145 rows x 2 columns]

<h3>Mednarodni prevozi</h3>



In [391]:
# International trips

sortedData['International'] = np.where(sortedData['Country - Origin'] == sortedData['Country - Destination'], True, False)

sortedData.groupby('International')['Weight'].agg(['sum','count'])

sum  count
International                   
False          2993133.83    359
True            230532.18     32

<h2>Zamude</h2>

<p>Poglejmo prevoze gleda na ETA status.</p>
<p>Spodaj so izpisani le primeri ko je ETA status različen od None. Takih primerov je le malo, ne moremo kaj več sklepati, podatki so nepopolni.</p>

In [392]:
#Zamude 
#ETA status : ['None', 'On time', 'Delayed']
statusData = sortedData.loc[sortedData['ETA status'] != 'None']
statusData

ID ETA status     ETA type Difference with ETA  Deliveries  Attachments  \
8    9    On time  Destination           -01:00:43           1            0   
9   10    On time  Destination           -00:09:14           1            0   
10  11    On time  Destination           -00:06:20           1            0   
12  13    On time  Destination           -00:15:03           1            0   
27  28    Delayed  Destination            00:00:58           1            1   

   City - Origin Country - Origin  Unique destinations City - Destination  \
8       Potpican               HR                    1               Oulx   
9       Potpican               HR                    1            Venzone   
10      Potpican               HR                    1            Mantova   
12      Potpican               HR                    1         Felsőlajos   
27      Potpican               HR                    1             Rimini   

   Country - Destination Vehicle (Requirement)    Weight Weight unit  \
8                     IT       Truck 100 m3 NO   6370.30          kg   
9                     IT       Truck 100 m3 NO   7668.83          kg   
10                    IT       Truck 100 m3 NO   8843.54          kg   
12                    HU       Truck 100 m3 NO  10754.83          kg   
27                    IT       Truck 100 m3 NO   5769.60          kg   

    Loading month  Loading day  Unloading month  Unloading day  \
8               8           30                9              1   
9               8           30                8             31   
10              8           30                8             31   
12              8           30                8             31   
27              8           26                8             29   

    Duration (days)  International  
8                 2          False  
9                 1          False  
10                1          False  
12                1          False  
27                3          False

<h2>Tipi Tovornjakov</h2>

<p> Vsota prevožene teže, števila prevozov in povprečne teže na vrsto tovornjaka.</p>

In [393]:
#Tipi tovornjakov

truckData = sortedData.groupby(['Vehicle (Requirement)'])['Weight'].agg(['sum','count']).sort_values(by=['count'])

truckData['Average weight'] = truckData['sum']/truckData['count']

truckData

sum  count  Average weight
Vehicle (Requirement)                                               
Truck 120 m3 NO                       9826.34      1     9826.340000
Megatrailer 95 m3                    28632.89      3     9544.296667
HU 24T one-way max 2 stops (5000)   124341.00      6    20723.500000
Truck 90 m3                          54175.60      8     6771.950000
Truck 100 m3 NO                    3006690.18    373     8060.831582

<p>Poglejmo kakšna je povezava med tipi tovornjakov in mednarodnimi prevozi.</p>

In [394]:
sortedData.groupby(['Vehicle (Requirement)', 'International'])['ID'].count()

Vehicle (Requirement)              International
HU 24T one-way max 2 stops (5000)  False              6
Megatrailer 95 m3                  False              3
Truck 100 m3 NO                    False            342
                                   True              31
Truck 120 m3 NO                    True               1
Truck 90 m3                        False              8
Name: ID, dtype: int64

<p>Povezava med tipi tovornjakov in začetno lokacijo</p>

In [395]:
sortedData.groupby(['Vehicle (Requirement)', 'City - Origin'])['ID'].count()

Vehicle (Requirement)              City - Origin
HU 24T one-way max 2 stops (5000)  HALMAJUGRA         6
Megatrailer 95 m3                  Caparroso          3
Truck 100 m3 NO                    Potpican         373
Truck 120 m3 NO                    Potpican           1
Truck 90 m3                        Potpican           8
Name: ID, dtype: int64

<h2>Zaključki</h2>

<ul>

<li>Imamo 391 vnosov podatkov iz obdobja 2022-05-31 do 2022-09-01.</li>
<li>Podjetje opravlja na prevoz maksimalno 3 dostave.</li>
<li>Prevoz se začne v treh mestih v treh različnih državah: Hr - Potpican, Es - Caparroso, Hu - Halmajugra.</li>
<li>Prevaža se v 7 različnih držav in 145 različnih mest.</li>
<li>Za različne dostave potrtebujejo 5 tipov tovornjakov.</li>
<li>9. in 5. mesec imata le po 5 in 2 vnosa prevozov. Ostali meseci so po številu prevozov in prevoženi teži dokaj enakomerno razporejeni (143, 143, 121).</li>
<li>Podjetje ima 98% prometa iz Hr - Potpican. Prevozi iz Caparrose so bili le 3 7. in 8. meseca. Iz Halmajugra so prevažali le 6-krat vse 8. meseca. </li>
<li>Podjetjej opravlja večinoma prevoze znotraj države 92%, le 8% je mednarodnih prevozov.</li>
<li>Končnih destinacij je 145, v najpogostejšo so dostavili 52-krat</li>
<li>Uporabljajo največ tovornjakov tipa Truck 100 m3 NO (95%).</li>
<li>Najbolj zmogljiv je tovornjak tipa HU 24T one-way max 2 stops (5000), ki je v povprečju prepeljal 20723 kg. </li>
<li>Za mednarodne prevoze se uporablje le tovornajke tipa Truck 100 m3 NO in Truck 120 m3 NO. </li>
<li>Na posamezni lokaciji se uporablja le en tip tovornjaka.</li>

</ul>

<h2>Opombe</h2>

<p>Pred analizo bi bilo dobro imeti malo več vsebine za podatke, določečeni stolpci niso povsem očitni glede pomena (npr. Attachments, je to število prikolic?). Tudi nekaj je nepopolnih vnosov (npr. ETA status). </p>

<p>Pri analizi se nisem preveč koncentriral na samo grafično predstavitev podatkov, ampak bolj na sam pomen podatkov.</p>

<p>Podatki skrivajo še marsikaj. Lahko bi gledali še kake podrobnosti poti z več postajami.</p> 

<p>Zanimivo bi bilo imeti nekoliko obsežnejše podatke, tako časovno kot bolj podrobno. Predvsem bi me zanimal ETA status, če bi imeli bolj natančen pregeld zamud in točnega časovnega trajanja poti, bi lahko iskali razloge za zamude (so nekatere poti vedno zamudne slaba ocena prihoda, se zamude pojavljajo samo z določenimi tovornjaki, na določene dneve tedna,...). Ali so imeli tovornjaki več postaj za oddajo tovora in katere se to vmesne postaje?</p>

<p>Naloga je bila zelo splošna, tako da sem Vam pripravil tudi bolj splošen pregled. Dobro bi bilo imeti kako tezo že podano, ki bi jo potem lahko potrdili ali ovrgli, ali pa imeti kak nadalnji cilj s danimi podatki (napovedovanje zamud, optimizacija poti).</p>

<p>Naloga je bila zanimiva in ne preveč globoka. Upam, da ste dobili željene informacije. Če bi želeli, da dodam še kaj mi sporočite. </p>

<p>:)</p>